### PR TITLE
Enumerable build

### PR DESCRIPTION
**Pull request by AshakaE**
The Enumerable module provides a collection of methods that get "mixed in", or included, with other classes like Array and Hash. The class must provide a method each, which yields successive members of the collection.

### CODE REVIEW

File: script.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
+module Enumerable
+  def my_each
+    return to_enum(:each) unless block_given?
+
+    arr = self if instance_of?(Array)
+    arr.length.times { |item| yield(arr[item]) }
+    self
+  end
```
Comment (javitocor) [inline script.rb]: #my_each when a block is given when self is a range returns the range itself after calling the 
given block once for each element in self
#my_each when a block is given when self is a range behaves like #each calls the given block once for each element in self
#my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self

File: script.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_each_with_index
+    return to_enum(:each_with_index) unless block_given?
+
+    arr = self if instance_of?(Array)
+    arr.length.times { |item| yield(arr[item], item) }
+    self
+  end
```
Comment (javitocor) [inline script.rb]: #my_each_with_index when a block is given when self is a range returns the range itself after 
calling the given block once for each element in self
#my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
#my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

File: script.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_select
+    new_arr = []
+    return to_enum(:select) unless block_given?
+
+    my_each { |m| new_arr << m if yield(m) }
+    new_arr
+  end
```
Comment (javitocor) [inline script.rb]: #my_select returns an array containing all elements of enum for which the given block returns 
a true value::range

File: script.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_all?
+    condition = true
+    my_each do |item|
+      condition = false if (block_given? && !yield(item)) || (!block_given? && !item)
+    end
+    condition
+  end
```
Comment (javitocor) [inline script.rb]: #my_all? returns false if the block returns false or nil::range
#my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
#my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
#my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
#my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
#my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
#my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
#my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

File: script.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_any?
+    condition = false
+    my_each do |item|
+      condition = true if (block_given? && yield(item)) || (!block_given? && item)
+    end
+    condition
+  end
```
Comment (javitocor) [inline script.rb]: #my_any? returns false if the block ever returns a value false or nil::range
#my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
#my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
#my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
#my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
#my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
#my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: script.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_none?
+    condition = true
+    my_each do |item|
+      condition = false if (block_given? && yield(item)) || (!block_given? && item)
+    end
+    condition
+  end
```
Comment (javitocor) [inline script.rb]: #my_none? returns false if the block ever returns true for all elements::range
#my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
#my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
#my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
#my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
#my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
#my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
#my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: script.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_count
+    return size unless block_given?
+
+    count = 0
+    my_each do |item|
+      count += 1 if yield(item)
+    end
+    count
+  end
```
Comment (javitocor) [inline script.rb]: #my_count counts the number of items in enum that are equal to item if an argument is given

File: script.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    new_map = []
+    arr.my_each do |item|
+      new_map << if block_given?
+                   yield(input.call(item))
+                 else
+                   input.call(item)
+                 end
+    end
+    new_map
+  end
```
Comment (javitocor) [inline script.rb]: #my_map returns a new array with the results of running a given block once for every element in enum.
#my_map returns a new array with the results of running a given block once for every element in enum.
#my_map returns a new array with the results of running a given block once for every element in enum::range
#my_map returns an Enumerator if no block is given
#my_map accepts a proc as an argument and returns a new array with the results of running the 
proc once for every element in enum.
#my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
#my_map does not mutate the original array

File: script.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      end
+    else
+      stack = self[0]
+      my_each_with_index do |_n, i|
+        stack = yield(stack, self[i + 1]) if i < length - 1
+      end
+    end
+    stack
+  end
+end
```
Comment (javitocor) [inline script.rb]: #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
#my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
#my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
#my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
#my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
#my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Hello team! 👋 
A great effort in this project 👍 
However, we still need to implement soma changes in order to get the project approved 💪 
Kindly check my comments below 👇  apply the changes and request a new code review ⏩ 
Good luck! 😃 
## Mandatory changes
- Add a license to your project
- Add an acknowledgements section to your readme file
- In your script.rb kindly remove all puts or examples, you can add them into a new file

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Good job working in my_each method
However, you still have to fix all the other methods and apply the mandatory changes
Kindly be sure to modify all of them before applying for a new code review. Do not forget to add the license and modify your readme file. Also, remember to remove all commented out code in your files
Good luck!

File: script.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
+module Enumerable
+  def my_each
+    return to_enum(:each) unless block_given?
+
+    # arr = self if instance_of?(Array)
```
Comment (ghost) [inline script.rb]: Remove the commented code on line 5.

File: script.rb [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    new_arr
+  end
+
+  def my_all?(_arg = nil)
+    condition = true
+    my_each do |item|
+      condition = false if _arg (block_given? && !yield(item)) || (!block_given? && !item)
+    end
+    condition
+  end
```
Comment (ghost) [inline script.rb]: The following test cases are failing for `my_all?` method.

```
- %w[ant bear cat].my_all? { |word| word.length >= 3 }
- %w[ant bear cat].my_all? { |word| word.length >= 4 }
- %w[ant bear cat].my_all?(/t/)
- [1, 2i, 3.14].my_all?(Numeric) 
- [nil, true, 99].my_all?  
- (0..5).my_all?
```

File: script.rb [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+  def my_any?(_arg = nil)
+    condition = false
+    my_each do |item|
+      condition = true if _arg (block_given? && yield(item)) || (!block_given? && item)
+    end
+    condition
+  end
```
Comment (ghost) [inline script.rb]: The following test cases are failing for `my_any?` method.

```
- %w[ant bear cat].my_any? { |word| word.length >= 4 }
- %w[ant bear cat].my_any?(/d/) 
- [nil, true, 99].my_any?(Integer) 
- [nil, true, 99].my_any? 
- (0..5).my_any?
```

File: script.rb [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+  def my_none?(_arg = nil)
+    condition = true
+    my_each do |item|
+      condition = false if _arg (block_given? && yield(item)) || (!block_given? && item)
+    end
+    condition
+  end
```
Comment (ghost) [inline script.rb]: Check your `my_none?` method against the following test cases.

```
- %w{ant bear cat}.my_none? { |word| word.length == 5 }
- %w{ant bear cat}.my_none? { |word| word.length >= 4 }
- %w{ant bear cat}.my_none?(/d/)  
- [1, 3.14, 42].my_none?(Float) 
- [nil].my_none? 
- [nil, false].my_none? 
- [nil, false, true].my_none?
- (0..4).my_none?
```

File: script.rb [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+  def my_count(value = nil, &block)
+    if block_given?
+      my_select(&block).size
+    elsif !value.nil?
+      my_select { |n| n == input }.size
+    else
+      size
+    end
+  end
```
Comment (ghost) [inline script.rb]: `my_count` method is failing the following test cases.

```
- [1, 2, 4, 2].my_count(2)
- (0..5).my_count(2)
```

File: script.rb [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    arr = to_a
+    arr.my_each do |item|
+      arr << if proc
+               proc.call(item)
+             else
+               yield(item)
+             end
+    end
+    arr
+  end
```
Comment (ghost) [inline script.rb]: `my_map` method is failing the following test cases

```
- [1,2,3,4,4,7,7,7,9].my_map { |i| i*4 }
- (1..4).my_map { |i| i*i }
- (1..4).my_map { "cat"  }
- double = Proc.new { |n| n * 2 }
 [1, 2, 3].my_map(&double) 
- my_map should execute only the proc when both a block and a proc are given. It should return a new array with the results of running the proc once for every element in enum. Example print [1, 2, 3].my_map(&double) { |i| i*i } should return [2,4,6].

```

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

## Hello team 👋 

**Great job implementing the requested changes 💪. However, there are a few fixes needed.**

1). **Add an acknowledgement section in your readme file. You can refer to the Microverse readme [template](https://github.com/microverseinc/readme-template).**

2). **You have a rubocop offense. see screenshot below.**
rubocop

3). **Check more in the comments below.**

## [Optional]
- A `.rubocop.yml` configuration file will help you do rubocop checks rather than doing it directly in your `script.rb` file.

## Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work, team. 💻 👀 ✨ 

- Excellent work passing most of the automated tests and fixing the rubocop errors.

## However, there are some changes required 👀 🧰 🔨 🔧 

- You have one last bug to fix. Here's the error message from the automated tester:
```
1) enumerables #my_map returns an Enumerator if no block is given
     Failure/Error: expect(array.my_map).to be_an(Enumerator)
     
     LocalJumpError:
       no block given (yield)
     # ./script.rb:71:in `block in my_map'
     # ./script.rb:6:in `block in my_each'
     # ./script.rb:6:in `times'
     # ./script.rb:6:in `my_each'
     # ./script.rb:67:in `my_map'
     # ./spec/enumerables_spec.rb:380:in `block (3 levels) in <top (required)>'
```

Just make sure that you return an `Enumerator` when no block is given to `my_map`.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by kevcap — APPROVED** [review]
# [STATUS] PROJECT APPROVED 🍾

### I really like your project! Congrats @AshakaE, you did an excellent job implementing the changes previously requested and you have my approval to go ahead and merge this pull request!



## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

**Pull request merged**